### ISSUE TITLE
Improve dtype mapping

### ISSUE THREAD

**Issue by yungyuc** [solvcon/solvcon#296]
In https://github.com/solvcon/modmesh/blob/44d4ccf8471678a1add2e830387bd02de348160a/cpp/modmesh/buffer/SimpleArray.cpp#L36 the list of if does not look like a perfect solution to the string to type code conversion:

```cpp
DataType get_data_type_from_string(const std::string & data_type_string)
{
    if (data_type_string == "bool")
    {
        return DataType::Bool;
    }
    if (data_type_string == "int8")
    {
        return DataType::Int8;
    }
    if (data_type_string == "int16")
    {
        return DataType::Int16;
    }
    if (data_type_string == "int32")
    {
        return DataType::Int32;
    }
    if (data_type_string == "int64")
    {
        return DataType::Uint64;
    }
    if (data_type_string == "uint8")
    {
        return DataType::Uint8;
    }
    if (data_type_string == "uint16")
    {
        return DataType::Uint16;
    }
    if (data_type_string == "uint32")
    {
        return DataType::Uint32;
    }
    if (data_type_string == "uint64")
    {
        return DataType::Uint64;
    }
    if (data_type_string == "float32")
    {
        return DataType::Float32;
    }
    if (data_type_string == "float64")
    {
        return DataType::Float64;
    }
    throw std::runtime_error("Unsupported datatype");
}
```

@visitorckw proposed to use a hash table for the mapping, like:

```cpp
DataType get_data_type_from_string(const std::string & data_type_string)
{
    const DataType table[4][4] = {
        {DataType::Bool, DataType::Bool, DataType::Bool, DataType::Bool},
        {DataType::Int32, DataType::Int64, DataType::Int16, DataType::Int8},
        {DataType::Uint32, DataType::Uint64, DataType::Uint16, DataType::Uint8},
        {DataType::Float32, DataType::Float64, DataType::Float32 DataType::Float64}
    };

    const char ch1 = data_type_string.front();
    const char ch2 = data_type_string.back();
    const int x = (ch1 - 'b');
    const int y = (ch2 - '0');
    const int i = (x & 3) ^ ((x & 4) >> 1);
    const int j = (y >> 1) - 1;

    return table[i & 3][j & 3];
}
```

The hash table should run faster than the branches.  The code is much more compact and should be more maintainable "in some ways".  The downside is obvious: the hash-table implementation is much more convoluted.  But it may be mitigated by proper comments and encapsulation.

A possible API could look like:

```cpp
class DataType
{
public:
    DataType(std::string const & name)
    {
        from_string(name);
    }
private:
    void from_string(std::string const & name)
    {
        // hash implementation.
    }
};
```

**Comment by visitorckw** [comment]
I'm not sure if the compiler is smart enough to further optimize this. We can change
```
const int y = (ch2 - '0');
const int j = (y >> 1) - 1;
```
to
```
const int y = (ch2 - '2');
const int j = (y >> 1);
```
to reduce one subtraction operation.

**Comment by yungyuc** [comment]
It might be more readable to write:

```cpp
const int y = ch2 - ('0' + 2);
const int j = y >> 1;
```

And it is interesting to see what is the assembly dictated by the compiler.

**Comment by tigercosmos** [comment]
we can make `DataType` also like enum. The practice can refer to this: https://stackoverflow.com/questions/21295935/can-a-c-enum-class-have-methods/53284026#53284026

```cpp
class DataType
{
public:
  enum Value : uint8_t
  {
    Bool,
    In8,
    Int16,
    ...
  };

  DataType() = default;
  constexpr DataType(Value datatype) : value(datatype) { }

#if Enable switch(datatype) use case:
  // Allow switch and comparisons.
  constexpr operator Value() const { return value; }

  // Prevent usage: if(datatype)
  explicit operator bool() const = delete;        
#else
  constexpr bool operator==(DataType a) const { return value == a.value; }
  constexpr bool operator!=(DataType a) const { return value != a.value; }
#endif

private:
  Value value;
};
```

**Comment by j8xixo12** [comment]
How about use a macro and lambda expression to implement a hash function to pre-calculate each data type string hash value in compile time.
Also this hash function also can be called in runtime and calculating the input string hash value `get_data_type_from_string()` then return the data type by a switch case statement.

The performance will be determined by the implementation of hash function.

I believe this proposal can both satisfy performance, readability and extendability.
```
// hash function macro, using lambda expression, let compiler decide it's return type in compile time
#define HASH_FUNC(str) [str] () { ...... }

// pre-calculate the each data type hash value in compile time
constexpr int uint8_hash = HASH_FUNC("uint8");
constexpr int uint16_hash = HASH_FUNC("uint16");
constexpr int uint32_hash = HASH_FUNC("uint32");

DataType get_data_type_from_string(const std::string & data_type_string)
{
    // Calculate input string hash value in runtime
    switch (HASH_FUNC(data_type_string))
    {
         case uint8_hash:
         return DataType::Uint8;
         break;

         case uint16_hash:
         return DataType::Uint16;
         break;

         case uint32_hash:
         return DataType::Uint32;
         break;
         
         default:
         break;
    }
    
    return DataType::NonType;
}
```

**Comment by tigercosmos** [comment]
based on @j8xixo12, we can use a hash table, and `unordered_map` already allows us to define customized hasher. The original concern is std::string default hasher may have a performance issue, but we can use a customized one instead.

**Comment by yungyuc** [comment]
@j8xixo12, lambda must incur function call overhead and it may be even slower than the if list in the current code base.  And like what @tigercosmos mentioned, hashing `std::string` is slow, too.  The branchless hashing proposed by @visitorckw looks much better, and the compact implementation might slightly improve maintainability, if my argument isn't overly flawed.

**Comment by tigercosmos** [comment]
based on the above conversation, I think the whole point is to have a fast hasher, so I have a similar approach in #297. @visitorckw  @yungyuc  would you like to give feedback of it?

**Comment by tigercosmos** [comment]
@yungyuc  we can close this.

**Comment by yungyuc** [comment]
This issue is resolved by #297 .
